tgchat

stylemistake opened the pull request:
## About The Pull Request

> Fixes #52898
> Fixes #52663

It has finally reached a full feature parity with goonchat, and is ready for review!

It is as fast as goonchat was (if not faster in certain circumstances), and is very extensible. It has all the necessary code for sorting messages into categories, which means that one of the next features will be multiple tab support.

Additional features that you will get with tgchat right now:

- Massively faster server-side performance compared to goonchat, especially if batching multiple messages to one client.
- Message persistence across rounds and reconnects. (All messages are stored client-side in IndexedDB)
- More robust scroll tracking. If you scroll up, it will not change the scroll position on new messages like goonchat did.
- Multiple message combining. (Currently set to combine up to 5 messages over last 5 seconds).
- If using the highlighting feature, it highlights the whole message as well as the matching word.
- "Now playing" widget, with preview of the song title, a knob for adjusting the volume and a stop button.

![image](https://user-images.githubusercontent.com/1516236/89689367-eb601900-d90c-11ea-938b-5daa67b28047.png)

---

Architecture is as following:

```
to_chat() -+
           |
        SSchat
   (queue, batching)
           |
  window.send_message()
           |
           v
+-------------+
| tgui-panel  |
|+-----------+|
|| tgchat    ||
|+-----------+|
+-------------+
```

Subsystem is basically `goonchat`, but without all the garbage that slows the servers down (string concatenation, double urlencoding, sanitizing, etc). Now, instead of all that, it's being slowed down by `json_encode` in `/datum/tgui_window/proc/send_message`, which IMO is completely worth it, and allows sending various templates and widgets to tgchat.

`/datum/tgui_window` abstracts the whole window away from you, establishes a nice message-passing interface between DM and JS, with two message queues on each side, automatically loads js/css assets for you, basically does everything. You as a developer only have to worry about sending/receiving messages and write javascript.

`tgui-panel` is a slimmed down version of `tgui`, and functions as a container for various widgets, and `tgchat` is one of them. It of course can be expanded with more stuff.

It's also a separate entry point and a JS bundle, so it's not bloating the main tgui bundle, and is currently sitting at about 230kB.

The grand plan is to replace the whole side panel, and have absolutely everything written in inferno, with mindblowing features and customizability.

stylemistake committed:
Initial panel setup (squashed)

Move stuff around to allow for codesplitting in the future.

Properly embedded chat.

Fix window.__augmentStack__

Some SSchat performance optimizations

- Proper message batching
- Use DOM fragments to insert batched messages
- No longer replace /proper /improper in TGUI_CREATE_MESSAGE

Disable runtime station mode

Revert skin.dmf

bobbahbrown commented:
i feel like you should remove the nanotrasen n from the background of the panel for chat but maybe thats just me

stylemistake commented:
Yeah, I will remove the NT logo, and will re-theme all tgui components to something more nice.

Jared-Fogle commented:
Is this going to force dark theme? I quite like the light theme for chat.

optimumtact commented:
finally light theme users will be executed

RaveRadbury commented:
![image](https://user-images.githubusercontent.com/3204033/88336062-33662580-cce9-11ea-86a0-ca9b556b5c2b.png)
@stylemistake being received by a grateful community after merging tgchat.

MrDoomBringer commented:
the start of an absolutely legendary feature
great work fam

stylemistake committed:
Implement ping, tweak styles, refactor

stylemistake committed:
New Tabs, message sorting and counting, improved ping

Nothing to see here

stylemistake committed:
Refactor panel, common chunk, retry loading stylesheets

stylemistake committed:
Theme switch, persist settings

Default font size to 9pt

stylemistake committed:
Rip out goonchat, move over missing bits

Fix the asset problem

stylemistake committed:
Use more guards, because tgui_panel might not always be available

stylemistake committed:
Implement telemetry

stylemistake committed:
Admin music playback

stylemistake committed:
Scroll tracking, improved music widget

saprasam commented:
make lightmode not shit
let me revert to the old shittier tgchat that i know and love

nemvar commented on code/modules/tgui_panel/tgui_panel.dm:
```diff
@@ -0,0 +1,128 @@
+
+/datum/tgui_panel/New(client/client)
+	src.client = client
+	window = new(client, "tgui-panel")
```
Maybe allow the panel to have an ID passed to it. It could come in handy when a second panel is required, for example for a tgui stat panel, unless I'm missing something.

https://github.com/tgstation/tgstation/pull/50963, for reference.

stylemistake replied:
The plan is to let that panel consume the whole right part of the screen, and in that case stat panel would be inside tgui-panel and stat subsystem would send messages to tgui-panel.

If it has to be a separate panel/window, you can do it with /datum/tgui_window, which is a general purpose abstraction. It does not require a lot of code to set up.

nemvar replied:
ah ok, thank you.

stylemistake committed:
IndexedDB, persistence, prune chat, client-side keypassing

stylemistake committed:
Fix gay baby jail

infoman-cor commented:
Icons do not appear to render correctly. They instead just show up as a missing image box.

Couls commented:
From the Testmerge, using ghostears and examining things when you have more than 1000 chat messages stored seems to freeze up the client and make it entirely unresponsive for a few seconds. Hasn't crashed my game yet but turning off ghostears stopped the freezing

infoman-cor commented:
I realize that this is still in its infancy but I would suggest adding string highlighting. As it really helps me know someone is attempting to talk to me in game.

ShibaInuLord commented:
Personally, I prefer the old style of the settings area just in the corner, and that having a whole bar with it saying chat and the amount of chat messages is kind of redundant.

stylemistake committed:
Bug fixes, state relay for external windows

Bunch of hotfixes

Fix problems with focus

Rebuild tgui

san7890 commented:
tgchat was test-merged today, and for some odd reason, the MOTD didn't play when I joined the server. When I clicked the "MOTD" verb, it played like normal, but it just didn't work when I joined the server for the first time after the test merge.

Wayland-Smithy commented:
>This pull request is still a work in progress

I am sure there will be improvements before a full merge, but I am having to play with it now so here is my feedback.

>for some odd reason, the MOTD didn't play when I joined the server.

During the TM I also noticed the missing automatic MOTD, and the apparent regression of the message highlighting feature 😢, and most of all the search (or possibly all text inputs) on other tgui windows became impossible for me to use.
![PW96JkK1pf](https://user-images.githubusercontent.com/64715958/89130994-8c3e7680-d4be-11ea-873f-b75d3f752103.gif)
At one point I got it to let me type enough to be useful, but I was not able to reproduce it reliably. If I got it to take any input and it was a movement hotkey it was also guaranteed to cause the "key stuck down" glitch, so it seems like a bug having to do with focus maybe?
![oU9cK1dQkt](https://user-images.githubusercontent.com/64715958/89131324-d58fc580-d4c0-11ea-871c-396822b55252.gif)
I am only clicking on each window once in this clip the rest of the focus changes are occurring on hover. I did confirm this does not happen on servers without tgchat.

Jared-Fogle commented:
It doesn't wrap long messages with no spaces.

TemporalOroboros commented:
tgchat was missing the 10+ minute afk timeout message when it disconnected me.

ishitbyabullet commented:
> When you use arrow keys in game, or arrow keys to scroll through tgchat, it seems to affect the other for some reason. For example, using the Up and Down arrow keys in tgchat seems to move my ghost character up and down as well. Vice versa the other way too, moving my ghost using arrow keys moves tgchat the same way.

Yeah, I have the same issue too.

Jared-Fogle commented:
During test merge, sometimes the X key would be hijacked and I can't switch hands with it. Unsure if this happens with others. Another player confirmed this.

Jared-Fogle commented:
It doesn't auto hyperlink.

Jared-Fogle commented:
BYOND supporter icon doesn't show.

stylemistake committed:
Game state notifications

stylemistake committed:
MIT License (JS)

stylemistake committed:
Fix the MOTD issue

stylemistake committed:
Move BYOND client theming to tgui-panel js

- Should in theory improve theme switching performance.

optimumtact commented:
Bug, when I press tab when chat is highlighted, instead of passing the command through to the game window, if I am near the bottom of chat it instead wraps around to the link at the top of the chat, scrolling my chat all the way back.

All key presses should be passed through to byond itself, and probably return focus to the map window.

Alternatives are to always return focus to the map window regularly, or on click (when that click doesn't hit an interactive element)

optimumtact commented:
also pressing f5 refreshes it and stops playing music

TemporalOroboros commented:
Potential bug, the "you have either been disconnected or are experiencing lag" popup never clears, even when the lag stops. Alternatively, it is entirely possible that my connectivity was just that bad that round.

Jared-Fogle commented:
Sometimes your movement will be forced in one direction. This has never happened so often without tgchat.

Jared-Fogle commented:
The x2 thing showing up for previous messages is annoying when using a health analyzer and switching between modes.

Wayland-Smithy commented:
From Basil with TM: 'tgchat' by stylemistake at commit 89afb7be95

The item icons when examining people are showing up as just dots like with byond supporter OOC:
![image](https://user-images.githubusercontent.com/64715958/89248577-f6d9da00-d5c4-11ea-98fd-1f609cef6653.png)

Sometimes some of them make it through. Most often it is just the top most, but sometimes it has no rhyme or reason as to which are dottified.
![image](https://user-images.githubusercontent.com/64715958/89248722-46200a80-d5c5-11ea-8f7e-c65acdfb5d2f.png)
![image](https://user-images.githubusercontent.com/64715958/89248869-9d25df80-d5c5-11ea-899b-b5ea6adf2d46.png)
![image](https://user-images.githubusercontent.com/64715958/89248992-ea09b600-d5c5-11ea-95c2-2bcc92789dc1.png)

optimumtact commented:
With regard to the icon issue, you probably need to do something like this
https://github.com/tgstation/tgstation/blob/b87f65d95e3e0a1315ec16ef68c0d5017a7ad8c9/code/modules/goonchat/browserassets/js/browserOutput.js#L238-L252

Because we cannot guarantee assets send by the asset system appear before they are rendered to chat, we have a reload handler that attempts to reload images several times before displaying them when it detects they errored.

THis was how MSO was able to replace bicon in #29690

You can see it being plugged into img tags in the full goonchat browser system.

optimumtact commented:
@Jared-Fogle it's very unlikely that bugs in ssinput are related to this PR

Jared-Fogle commented:
@optimumtact Yeah, I know the connection is shaky, but anecdotal evidence is that a lot of people have suddenly been reporting this ever since the test merge.

san7890 commented:
> @optimumtact Yeah, I know the connection is shaky, but anecdotal evidence is that a lot of people have suddenly been reporting this ever since the test merge. I haven't played a round with it off yet for a while though, I will report if it's un-TM'd and it still happens to anyone I know.

The issues I reported above included how arrow keys (used to scroll through chat) would impact the character in game. Sometimes, using the arrow keys in game would also impact how chat scrolled. I believe this is the key reason why people's movement gets "stuck" until they reconnect or restart BYOND. This can be replicable by going into any server with this iteration of tgchat active and trying to use the arrow keys to scroll up and down in chat.

I'm not sure _why_ movement-sticks the way it happens, but arrow key movement and tgchat are definitely related.

Jared-Fogle commented:
I have gotten a report that the movement bug has been replicated on Sybil, which does not have tgchat test merged. I will file a separate issue for it. @optimumtact @san7890

optimumtact commented:
The issue occurs because a key gets stuck in the users keydown list, even when it's not actively being pressed. There's no known cause at this stage and I can't get it to occur reliably.

ShizCalev commented:
Minimum font size seems to be locked at 12px for me?

stylemistake commented:
Yes, why would you want to make it smaller?

uomo91 commented:
Please try to re-implement highlighting.

stylemistake committed:
Experimental focus hacks

ShizCalev commented:
@stylemistake What do you mean why would I want to make font smaller than 12px lol? 

My primary monitor is a 48" 1080p display; so what's normally tiny font for people using a 14" laptop display is instead -massive- on my screen, and thus I keep my text at a proportional size. The difference between 12px and 11px alone is seeing 6 additional lines of text without even factoring in message length making you see less due to wrapping as well. 

Please maintain current chat functionality by allowing you to set the text smaller. ♥

stylemistake commented:
I'll make it go as low as 8px, anything smaller than that may make it impossible to turn it back because the number input will become infinitely small.

stylemistake committed:
Message highlighting

stylemistake committed:
Fix connection restored notification

stylemistake committed:
Retry loading images in the chat

stylemistake committed:
Linkify, fix openLink, fix popup focus

stylemistake committed:
Fix middleware init, remove sync_chat_regexes

Rohesie commented on code/__DEFINES/tgui.dm:
```diff
@@ -26,3 +26,10 @@
 #define TGUI_WINDOW_INDEX(window_id) text2num(copytext(window_id, 13))
+
+/// Creates a message packet for sending via output()
+#define TGUI_CREATE_MESSAGE(type, payload) ( \
+	url_encode(json_encode(list( \
+		"type" = type, \
+		"payload" = payload, \
+	))))
```
```suggestion
#define TGUI_CREATE_MESSAGE(type, payload) do( \
	url_encode(json_encode(list( \
		"type" = type, \
		"payload" = payload, \
	)))) while(FALSE)
```

stylemistake replied:
Didn't compile.

Couls replied:
```suggestion
#define TGUI_CREATE_MESSAGE(type, payload) do {( \
	url_encode(json_encode(list( \
		"type" = type, \
		"payload" = payload, \
	)))) } while(FALSE)
```

stylemistake committed:
Clean up

Remove benchmark procs

Use a canonical winset form in focusMap

stylemistake committed:
Chat tabs, unread counts

BUGS!

One more

Fix conflicts

TemporalOroboros commented:
BUG: Currently the "remove chat tab" button in the tab settings, when clicked, removes every chat tab _except_ the one currently selected.

stylemistake commented:
> BUG: Currently the "remove chat tab" button in the tab settings, when clicked, removes every chat tab except the one currently selected.

Already fixed.

Jared-Fogle commented:
It shouldn't be possible to remove all tabs. It also shouldn't be possible to have no tabs with admin PMs on.

stylemistake commented:
Admin pm's are not categorized right now as such. Feel free to make a PR on my branch with improved filtering because i am going to be away for a few days.

LemonInTheDark commented:
I love these chat tabs so much

CoffeeDragon16 commented:
This is a MASSIVE improvement, best in a good while. It's just very good. I love the tab feature. 👍 Good job

spessbro commented:
doesn't work under wine sadly ;-;

Jared-Fogle commented:
It'd be nice if word highlighting didn't highlight for names. I want to know when people say my name, but I don't want every message with my name in it highlighted. Goonchat did the same, and I didn't like it there either.

TheSilverNuke commented:
With the new chat tab sorting system, I'd suggest moving PDA messages under Radio, that way we can avoid frivolous information such as this:

![2020-08-09 04_26_25-_tg_Station Basil  ENGLISH   US-WEST   100% FREE LAG _ Construct your own ship s](https://user-images.githubusercontent.com/36066032/89730054-8c77ce00-d9f8-11ea-8a56-4cf95728ac8e.png)

stylemistake committed:
IE8 compat

stylemistake committed:
Fix asset code

- tgui-dev-server works again
- Looks up and notifies dreamseeker to reload tgui windows

stylemistake committed:
Chat log saving

stylemistake committed:
Load byond macros

dondonpow commented:
Not sure if this is a test merge any more, but I'm having some issues. Whenever I try to connect, if the tgchat is used my client instantly crashes. This first started when I was trying to change a highlighted string, and when I tried to apply it, it crashed.

stylemistake commented:
@dondonpow I need more info, you can pm be in Discord, `stylemistake#1991`.

stylemistake committed:
Fix highlight regex crashing the client

stylemistake commented:
It's ready for merging (doubt anyone will review it). Most of the bugs were squashed and it was testmerged for a full day with no errors in the logs, so it's safe to merge.

Begin the new era of /tg/station now!

stylemistake committed:
Fix opening of chat tab settings on adding a tab

TheSilverNuke commented:
When filtering by dead chat the ghosts become unfollowable.
![2020-08-14 05_03_22-_tg_Station Basil  ENGLISH   US-WEST   100% FREE LAG _ Rogue Edition Quebec](https://user-images.githubusercontent.com/36066032/90243131-8ebc9c80-ddeb-11ea-914e-c0a010115fae.png)
![2020-08-14 05_03_10-_tg_Station Basil  ENGLISH   US-WEST   100% FREE LAG _ Rogue Edition Quebec](https://user-images.githubusercontent.com/36066032/90243134-8f553300-ddeb-11ea-8a9d-58ace17acd83.png)
